sqlalchemy-stubs/sql/type_api.pyi

```diff
@@ -30,67 +35,85 @@ class TypeEngine(Visitable, Generic[_T_co]):
-    def bind_processor(self, dialect): ...
-    def result_processor(self, dialect, coltype): ...
-    def column_expression(self, colexpr): ...
-    def bind_expression(self, bindvalue): ...
-    def compare_values(self, x, y): ...
-    def get_dbapi_type(self, dbapi): ...
+    def evaluates_none(self: _TE) -> _TE: ...
+    def copy(self: _TE, **kw: Any) -> _TE: ...
+    def compare_against_backend(self, dialect: Dialect, conn_type: Any) -> Any: ...
+    def copy_value(self, value: _T) -> _T: ...
+    def literal_processor(self, dialect: Dialect) -> Optional[Callable[..., Any]]: ...
+    def bind_processor(self, dialect: Dialect) -> Optional[Callable[..., Any]]: ...
+    def result_processor(self, dialect: Dialect, coltype: Any) -> Optional[Callable[..., Any]]: ...
+    def column_expression(self, colexpr: Union[str, ColumnElement[Any]]) -> Optional[ColumnElement[Any]]: ...
+    def bind_expression(self, bindvalue: BindParameter[Any]) -> Optional[Any]: ...
+    def compare_values(self, x: Any, y: Any) -> bool: ...
+    def get_dbapi_type(self, dbapi: Any) -> Any: ...
     @property
     def python_type(self) -> Type[_T_co]: ...
-    def with_variant(self, type_, dialect_name): ...
-    def dialect_impl(self, dialect): ...
-    def adapt(self, *args, **kw): ...
-    def coerce_compared_value(self, op, value): ...
-    def compile(self, dialect: Optional[Any] = ...): ...
+    def with_variant(self, type_: TypeEngine[_U], dialect_name: str) -> Variant[_U]: ...
+    def dialect_impl(self, dialect: Dialect): ...
+    def adapt(self, cls: Type[_U], **kw: Any) -> _U: ...
+    def coerce_compared_value(self, op: Any, value: Any) -> TypeEngine[Any]: ...
+    def compile(self, dialect: Optional[Dialect] = ...): ...
 
 class VisitableCheckKWArg(util.EnsureKWArgType, VisitableType): ...
 
+_UDT = TypeVar('_UDT', bound=UserDefinedType)
+
 class UserDefinedType(TypeEngine[_T], metaclass=VisitableCheckKWArg):
     __visit_name__: str = ...
     ensure_kwarg: str = ...
     class Comparator(TypeEngine.Comparator): ...
     comparator_factory: Any = ...
-    def coerce_compared_value(self, op, value): ...
+    def coerce_compared_value(self: _UDT, op: Any, value: Any) -> _UDT: ...
+
+class Emulated(object):
+    def adapt_to_emulated(self, impltype: Type[Any], **kw: Any) -> Any: ...
+    def adapt(self, cls: Type[Any], **kw: Any) -> Any: ...
+
+_NFE = TypeVar('_NFE', bound=NativeForEmulated)
+
+class NativeForEmulated(object):
+    @classmethod
+    def adapt_emulated_to_native(cls: Type[_NFE], impl: Any, **kw: Any) -> _NFE: ...
+
+_TD = TypeVar('_TD', bound=TypeDecorator)
 
 class TypeDecorator(SchemaEventTarget, TypeEngine[_T]):
     __visit_name__: str = ...
-    impl: Type[TypeEngine[Any]] = ...  # TODO: second type variable? (return of process_bind_param)
-    def __init__(self, *args, **kwargs) -> None: ...
-    coerce_to_is_types: Any = ...
+    impl: TypeEngine[Any] = ...  # TODO: second type variable? (return of process_bind_param)
```

Comment:
The issue here is that on the class it's `Type[TypeEngine[Any]]`, but on the instance, it's `TypeEngine[Any]` because of this code in `TypeDecorator.__init__()`:

```python
        self.impl = to_instance(self.__class__.impl, *args, **kwargs)
```

Typing this as a union is going to lead to unnecessary checks for instances. Should we just set it to `Any` for now?

Reply:
OK, lets leave all these `Any` and leave a comment why.